add optional doughnut parameter to contract calls

cowboy-bebug opened the pull request:
Changes:
- [x] ExecutionContext has optional doughnut field
    - added optional doughnut param to methods where needed
    - for tests and node/runtime, None is used as doughnut for now
- [x]  Expose execution context doughnut via new doughnut() method on the Ext trait
- [x] Add contract-to-contract hook using Ext::doughnut()

Closes #33

cowboy-bebug committed:
add optional doughnut parameter to contract calls

- ExecutionContext has optinal doughnut associated type
- added optional doughnut param to methods where needed
- for tests, None is used as doughnut for now

cowboy-bebug committed:
add doughnut method to Ext trait and ext_doughnut wasm function

jordy25519 commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -903,6 +909,49 @@ mod tests {
+		(call $assert
+			(i32.eq
+				(call $ext_doughnut)
+				(i32.const 0)
```
This is constant `0`, comment above says `1`

jordy25519 commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -903,6 +909,49 @@ mod tests {
 		).unwrap();
 	}
 
+	/// calls `ext_doughnut`, loads the doughnut from the scratch buffer and
```
This test is only checking the return status code: it should assert the contents of the buffer aswell, like with this one: https://github.com/plugblockchain/plug-blockchain/blob/37c0fcfb34923ce5692932b76fe97b5f7cb68580/frame/contracts/src/wasm/mod.rs#L880-L895

jordy25519 commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -560,6 +560,18 @@ define_env!(Env, <E: Ext>,
 
+	// Stores the doughnut of the current contract into the scratch buffer.
+	// Returns 0 if doughnut is present, otherwise 1.
+	ext_doughnut(ctx) -> u32 => {
```
We don't need this function to be implemented yet, if at all

cowboy-bebug replied:
Thanks! I'll remove it and the related test 😄

jordy25519 commented on frame/contracts/src/exec.rs:
```diff
@@ -284,6 +288,7 @@ pub struct ExecutionContext<'a, T: Trait + 'a, V, L> {
 	pub timestamp: MomentOf<T>,
 	pub block_number: T::BlockNumber,
 	pub origin: T::AccountId,
+	pub doughnut: Option<T::Doughnut>,
```
We should make this by ref if possible `Option<&T::Doughnut>` and avoid all the clones

jordy25519 reviewed (changes requested):
PR looks good in general
- Don't need `ext_doughnut` call to be implemented for this PR but left feedback anyway ;)
- Pass the doughnut by ref if possible (may require change to @alexsednz s work)

cowboy-bebug committed:
remove ext_doughnut and related test as not needed

cowboy-bebug committed:
use doughnut by ref to avoid clone

cowboy-bebug committed:
fix format

aliXsed commented on frame/contracts/src/exec.rs:
```diff
@@ -284,6 +288,7 @@ pub struct ExecutionContext<'a, T: Trait + 'a, V, L> {
 	pub timestamp: MomentOf<T>,
 	pub block_number: T::BlockNumber,
 	pub origin: T::AccountId,
+	pub doughnut: Option<&'a T::Doughnut>,
```
Wouldn't it be better that we pass the doughnut in through RawOrigin::Signed(issuer, doughnut) as it is for RawOrigin::Delegated for the sake of consistency?

cowboy-bebug replied:
Yeah I wasn't sure which doughnut to use so just used one from `system::Trait`

aliXsed replied:
Yup, if we do so then we wouldn't have two doughnuts passed in directly and indirectly through RawOrigin.

cowboy-bebug replied:
I see what u mean but they are the same thing as `system::Trait` is defined as:
```rust
pub trait Trait: 'static + Eq + Clone {
    /// The aggregated `Origin` type used by dispatchable calls.
    type Origin:
        Into<Result<RawOrigin<Self::AccountId, Self::Doughnut>, Self::Origin>> + 
        From<RawOrigin<Self::AccountId, Self::Doughnut>> +
    ...
}
```
`Self::Doughnut` inside Origin is the `T::Doughnut`

aliXsed replied:
That's why I suggest we should only have one of them. If they could be different then there could be scenarios where we would need to get either of these doughnuts separately and that could justify passing in two references. Does it make sense?

cowboy-bebug replied:
It's more clear and consistent to me that we are using `system::Trait::Doughnut` type, similar to how RawOrigin and DelegatedDispatchVerifier are also using `system::Trait::Doughnut` type. And we are not passing anything just yet, it's just declaring what the type should be for the doughnut field.

cowboy-bebug committed:
verify doughnut, if attached, in contract to contract calls

jordy25519 commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -387,6 +388,15 @@ define_env!(Env, <E: Ext>,
+		if let Some(doughnut) = ctx.ext.doughnut() {
+			type DelegatedDispatchVerifierOf<E> = <<E as Ext>::T as system::Trait>::DelegatedDispatchVerifier;
+			DelegatedDispatchVerifierOf::<E>::verify_contract_to_contract_call(
+				&callee,
```
This one should be the `origin` address

jordy25519 commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -387,6 +388,15 @@ define_env!(Env, <E: Ext>,
 			read_sandbox_memory_as(ctx, value_ptr, value_len)?;
 
+		if let Some(doughnut) = ctx.ext.doughnut() {
+			type DelegatedDispatchVerifierOf<E> = <<E as Ext>::T as system::Trait>::DelegatedDispatchVerifier;
```
Can move the type alias to the top of the file

hoani reviewed (approved):
LGTM

jordy25519 reviewed (changes requested):
Looks good! Should add a couple integration tests to check `verify_contract_to_contract_call` works correctly. forgot to request this for `runtime_to_contract` in #47

cowboy-bebug committed:
use origin to verify contract to contract call

cowboy-bebug committed:
add doughnut integration tests for contract to contract call
- test setup is the same as the contracts module (tests.rs)
- the tests are inspired by deploy_and_call_other_contract() in tests.rs

jordy25519 commented:
LGTM 🎉 🎉 🎉

jordy25519 commented on frame/contracts/src/lib.rs:
```diff
@@ -572,12 +574,13 @@ decl_module! {
 			#[compact] gas_limit: Gas,
-			data: Vec<u8>
+			data: Vec<u8>,
+			doughnut: Option<T::Doughnut>,
```
so `doughnut` should be extracted from the `origin`. the additional parameter here and in `instantiate` should be removed

cowboy-bebug replied:
Fixed by https://github.com/plugblockchain/plug-blockchain/pull/49
